### PR TITLE
fix(locksmith): better handling of timeouts when retrieving data from subgraphs

### PR DESCRIPTION
**Pull request by julien51**
# Description

As I was investigating why locksmith was serving some 500 from time to time I realized we are not properly handling timeouts when retrieving data from the subgraphs. This PR adds the ability to timeout "earlier".

I also found that we are making 2 requests to subgraphs (one to get the expiration and one to get the lock name) which is very wasteful. We will have to clean that up in another PR.

# Checklist:

- [X] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: locksmith/src/utils/promises.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+export const waitOrTimeout = async (
```
Comment (julien51) [inline locksmith/src/utils/promises.ts:1]: A simple util to race with a timeout and yield back control to the caller when that happens.

File: locksmith/src/utils/promises.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  promise: Promise<any>,
+  delay: number,
+  responseIfTimeout: () => any
```
Comment (clemsos) [inline locksmith/src/utils/promises.ts:4]: Do these args really need to be passed? `delay` could have directly `TIMEOUT_ON_SUBGRAPH` set as default value, then just a smaller error message string

File: locksmith/src/utils/metadata.ts [hunk 3]
```diff
@@ -1,23 +1,45 @@
...
-      {
-        network,
+        {
+          network,
+        }
+      ),
+      TIMEOUT_ON_SUBGRAPH,
+      () => {
+        logger.error(
+          `Timed out after 3s while retrieving info from Subgraph for lock metadata ${lockAddress} on ${network}`
```
Comment (clemsos) [inline locksmith/src/utils/metadata.ts:39]: the `3s` should be parsed from `TIMEOUT_ON_SUBGRAPH` for consistency

**Review by clemsos — APPROVED** [review]
few comments for simplification but lg otherwise

File: locksmith/src/utils/metadata.ts (hunk 3 above)
Comment (julien51) [reply locksmith/src/utils/metadata.ts:39]: Great point! I'll do that!

**Pull request merged**